presto-common/src/main/java/com/facebook/presto/common/block/ColumnarArray.java

```diff
@@ -0,0 +1,207 @@
+                0,
+                offsets,
+                new DictionaryBlock(dictionaryIds.length, elementsBlock, dictionaryIds),
+                INSTANCE_SIZE + dictionaryBlock.getRetainedSizeInBytes() + sizeOf(offsets) + sizeOf(dictionaryIds),
+                // The estimated serialized size is the sum of the following
+                // 1) the offsets size: Integer.BYTES * positionCount
+                // 2) nulls array size: Byte.BYTES * positionCount
+                // 3) the estimated serialized size for the elementsBlock which was just constructed as a new DictionaryBlock:
+                //     the average row size: elementsBlock.getSizeInBytes() / (double) elementsBlock.getPositionCount() * the number of rows: offsets[positionCount]
+                (Integer.BYTES + Byte.BYTES) * positionCount + elementsBlock.getPositionCount() == 0 ? 0 : (long) (elementsBlock.getSizeInBytes() / (double) elementsBlock.getPositionCount() * offsets[positionCount]));
+    }
+
+    private static ColumnarArray toColumnarArray(RunLengthEncodedBlock rleBlock)
+    {
+        ColumnarArray columnarArray = toColumnarArray(rleBlock.getValue());
+        int positionCount = rleBlock.getPositionCount();
+
+        // build new offsets block
+        int[] offsets = new int[positionCount + 1];
+        int valueLength = columnarArray.getLength(0);
+        for (int i = 0; i < offsets.length; i++) {
+            offsets[i] = i * valueLength;
+        }
+
+        // create indexes for a dictionary block of the elements
+        int[] dictionaryIds = new int[positionCount * valueLength];
+        int nextDictionaryIndex = 0;
+        for (int position = 0; position < positionCount; position++) {
+            for (int entryIndex = 0; entryIndex < valueLength; entryIndex++) {
+                dictionaryIds[nextDictionaryIndex] = entryIndex;
+                nextDictionaryIndex++;
+            }
+        }
+
+        Block elementsBlock = columnarArray.getElementsBlock();
+        return new ColumnarArray(
+                rleBlock,
+                0,
+                offsets,
+                new DictionaryBlock(dictionaryIds.length, elementsBlock, dictionaryIds),
+                INSTANCE_SIZE + rleBlock.getRetainedSizeInBytes() + sizeOf(offsets) + sizeOf(dictionaryIds),
+                // The estimated serialized size is the sum of the following
+                // 1) the offsets size: Integer.BYTES * positionCount
+                // 2) nulls array size: Byte.BYTES * positionCount
+                // 3) the estimated serialized size for the elementsBlock which was just constructed as a new DictionaryBlock:
+                //     the average row size: elementsBlock.getSizeInBytes() / (double) elementsBlock.getPositionCount() * the number of rows: offsets[positionCount]
+                (Integer.BYTES + Byte.BYTES) * positionCount + (long) (elementsBlock.getSizeInBytes() / (double) elementsBlock.getPositionCount() * offsets[positionCount]));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).